File: bindings/python/demos/animated-marker.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#This is a demonstration of the libchamplain python binding.
#It will display a blue "pulsating" dot on a given position on the map
#A old legend that this position is the bed of Pierlux. You will have to find out...

import champlain
import clutter
import cairo
import math
import gobject

MARKER_SIZE = 10
MARKER_COLOR = [0.1,0.1,0.9,1.0]

POSITION = [45.528178, -73.563788]
SCREEN_SIZE = [640, 480]

#The AnimatedMarker will extend the champlain.Marker class
class AnimatedMarker(champlain.Marker) :
    def __init__(self,color=None) :
        champlain.Marker.__init__(self)
        
        orange = clutter.Color(0xf3, 0x94, 0x07, 0xbb)
        white = clutter.Color(0xff, 0xff, 0xff, 0xff)
        if not color :
            color = MARKER_COLOR
        
        #Cairo definition of the inner marker
        bg_in = clutter.CairoTexture(MARKER_SIZE, MARKER_SIZE)
        cr_in = bg_in.cairo_create()
        cr_in.set_source_rgb(0, 0, 0)
        cr_in.arc(MARKER_SIZE / 2.0, MARKER_SIZE / 2.0, MARKER_SIZE / 2.0, 0, 2 * math.pi)
        cr_in.close_path()
        cr_in.set_source_rgba(*color)
        cr_in.fill()
        self.add(bg_in)
        bg_in.set_anchor_point_from_gravity(clutter.GRAVITY_CENTER)
        bg_in.set_position(0, 0)
        
        #Cairo definition of the outside circle (that will be animated)
        bg_out = clutter.CairoTexture(2 * MARKER_SIZE, 2 * MARKER_SIZE)
        cr_out = bg_out.cairo_create()
        cr_out.set_source_rgb(0, 0, 0)
        cr_out.arc(MARKER_SIZE, MARKER_SIZE, 0.9 * MARKER_SIZE, 0, 2 * math.pi)
        cr_out.close_path()
        cr_out.set_line_width(2.0)
        cr_out.set_source_rgba(*color)
        cr_out.stroke()
        self.add(bg_out)
        bg_out.lower_bottom()
        bg_out.set_position(0, 0)
        bg_out.set_anchor_point_from_gravity(clutter.GRAVITY_CENTER)

        #The timeline of our animation
        self.timeline = clutter.Timeline()
        self.timeline.set_duration(1000)
        self.timeline.set_loop(True)

        self.alpha = clutter.Alpha(self.timeline, clutter.EASE_OUT_SINE)
        
        #The "growing" behaviour
        self.grow_behaviour = clutter.BehaviourScale(0.5, 0.5, 2.0, 2.0)
        self.grow_behaviour.set_alpha(self.alpha)
        self.grow_behaviour.apply(bg_out)

        #The fade out behaviour
        self.fade_behaviour = clutter.BehaviourOpacity(255, 0)
        self.fade_behaviour.set_alpha(self.alpha)
        self.fade_behaviour.apply(bg_out)

        #If you want the marked to be animated without calling start_animation
        #Uncomment the following line
        #self.timeline.start()
        
    def stop_animation(self) :
        self.timeline.stop()
        
    def start_animation(self) :
        self.timeline.start()
        

def main() :
    gobject.threads_init()
    clutter.init()
    stage = clutter.stage_get_default()
    actor = champlain.View()
    layer = champlain.Layer()
    marker = AnimatedMarker()
    #Starting the animation, that's what we want after all !
    marker.start_animation()
    #Set marker position on the map
    marker.set_position(*POSITION)
    
    stage.set_size(*SCREEN_SIZE)
    actor.set_size(*SCREEN_SIZE)
    stage.add(actor)
    
    layer.show()
    actor.add_layer(layer)
    layer.add(marker)

    # Finish initialising the map view
    actor.set_property("zoom-level", 5)
    actor.set_property("scroll-mode", champlain.SCROLL_MODE_KINETIC)
    actor.center_on(*POSITION)
    
    stage.show()
    
    clutter.main()
        

if __name__ == "__main__":
    main()

File: bindings/perl/Champlain/examples/animated-marker.pl
#!/usr/bin/perl

use strict;
use warnings;

use Glib qw(TRUE FALSE);
use Clutter qw(-threads-init -init);
use Champlain;
use Math::Trig ':pi';


my $MARKER_SIZE = 10;


exit main();


sub main {
	
	my $stage = Clutter::Stage->get_default();
	$stage->set_size(800, 600);
	
	# Create the map view
	my $map = Champlain::View->new();
	$map->set_size($stage->get_size);
	$stage->add($map);
	
	
	# Create the marker layer
	my $layer = Champlain::Layer->new();
	$map->add_layer($layer);
	
	# Create the marker
	my $marker = create_marker();
	$layer->add($marker);
	
	
	# Finish initializing the map view
	$map->set_property("zoom-level", 5);
	$map->set_property("scroll-mode", 'kinetic');
	$map->center_on(45.466, -73.75);
	

	$stage->show_all();
	
	Clutter->main();
	
	return 0;
}


#
# The marker is drawn with cairo. It is composed of 1 static filled circle and 1
# stroked circle animated as an echo.
#
sub create_marker {
	my $marker = Champlain::Marker->new();
	
	# Static filled circle
	create_static_circle($marker);
	
	
	# Echo circle
	my $echo = create_echo_circle($marker);
	
	##
	## Animate the echo circle
	##
	my $timeline = Clutter::Timeline->new_for_duration(1000);
	$timeline->set_loop(TRUE);
	my $alpha = Clutter::Alpha->new($timeline, \&Clutter::Alpha::sine_inc);
	
	my $behaviour = Clutter::Behaviour::Scale->new($alpha, 0.5, 0.5, 2.0, 2.0);
	$behaviour->apply($echo);
	
	$behaviour = Clutter::Behaviour::Opacity->new($alpha, 255, 0);
	$behaviour->apply($echo);
	$timeline->signal_connect('new-frame', sub {
		my ($timeline, $frame) = @_;
		print $frame, "\n";
	});
	$timeline->start();
	
	
	# Sets marker position on the map
	$marker->set_position(45.528178, -73.563788);
	
	return $marker;
}


sub create_static_circle {
	my ($marker) = @_;
	
	my $texture = Clutter::Texture::Cairo->new($MARKER_SIZE, $MARKER_SIZE);
	my $cr = $texture->create_context();
	
	# Draw the circle
	$cr->set_source_rgb(0, 0, 0);
	$cr->arc($MARKER_SIZE/2, $MARKER_SIZE/2, $MARKER_SIZE/2, 0, pi2);
	$cr->close_path();

	# Fill the circle
	$cr->set_source_rgba(0.1, 0.1, 0.9, 1.0);
	$cr->fill();
	
	# Add the circle to the marker
	$marker->add($texture);
	$texture->set_anchor_point_from_gravity('center');
	$texture->set_position(0, 0);
	
	return $texture;
}


sub create_echo_circle {
	my ($marker) = @_;
	
	my $texture = Clutter::Texture::Cairo->new($MARKER_SIZE * 2, $MARKER_SIZE * 2);
	my $cr = $texture->create_context();
	
	# Draw the circle
	$cr->set_source_rgb(0, 0, 0);
	$cr->arc($MARKER_SIZE, $MARKER_SIZE, $MARKER_SIZE * 0.9, 0, pi2);
	$cr->close_path();

	# Stroke the circle
	$cr->set_line_width(2.0);
	$cr->set_source_rgba(0.1, 0.1, 0.7, 1.0);
	$cr->stroke();

	# Add the circle to the marker
	$marker->add($texture);
	$texture->lower_bottom(); # Ensure it is under the previous circle
	$texture->set_position(0, 0);
	$texture->set_anchor_point_from_gravity('center');
	
	return $texture;
}
